subsys/lorawan/nvm/lorawan_nvm_settings.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2022 Intellinium <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <LoRaMac.h>
+#include <kernel.h>
+#include <settings/settings.h>
+#include "lorawan_nvm.h"
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(lorawan_nvm, CONFIG_LORAWAN_LOG_LEVEL);
+
+static uint16_t nvm_notify_flag;
+
+#define LORAWAN_SETTINGS_BASE                "lorawan/nvm"
+
+#define flag_set(f)        ((nvm_notify_flag & (f)) == (f))
+
+#define SAVE_SETTINGS(nvm, flag, name)					  \
+if (flag_set(flag)) {							  \
+	LOG_DBG("Saving configuration " LORAWAN_SETTINGS_BASE "/" # name);\
+	int err = settings_save_one(LORAWAN_SETTINGS_BASE "/" # name,	  \
+				&(nvm)->name,				  \
+				sizeof((nvm)->name));			  \
+	if (err) {							  \
+		LOG_ERR("Could not save settings" # name", error %d",	  \
+			err);						  \
+	}								  \
+}
+
+static void lorawan_nvm_save_settings(void)
+{
+	MibRequestConfirm_t mibReq;
+
+	LOG_DBG("Saving LoRaWAN settings");
+
+	/* Retrieve the actual context */
+	mibReq.Type = MIB_NVM_CTXS;
+	if (LoRaMacMibGetRequestConfirm(&mibReq) != LORAMAC_STATUS_OK) {
+		LOG_ERR("Could not get NVM context");
+		return;
+	}
+
+	LoRaMacNvmData_t *nvm = mibReq.Param.Contexts;
+
+	LOG_DBG("Crypto version: %"PRIu32", DevNonce: %d, JoinNonce: %"PRIu32,
+		mibReq.Param.Contexts->Crypto.LrWanVersion.Value,
+		mibReq.Param.Contexts->Crypto.DevNonce,
+		mibReq.Param.Contexts->Crypto.JoinNonce);
+
+	/* Wait for LoRa to be idle */
+	while (LoRaMacStop() != LORAMAC_STATUS_OK) {
```

Comment:
As spotted by @JordanYates, we can safely avoid stopping and starting the LoRaWAN stack here.

Reply:
done